ddl/index.go

```diff
@@ -386,111 +396,90 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) (ver int64, _ error) {
-	}
-	// Be sure to do this operation only once.
-	if !handleInfo.isSent {
-		// Notice to start the next task operation.
-		taskOpInfo.nextCh <- ret.doneHandle
-		// Record the last handle.
-		// Ensure that the handle scope of the task doesn't change,
-		// even if the transaction retries it can't effect the other tasks.
-		handleInfo.endHandle = ret.doneHandle
-		handleInfo.isSent = true
-	}
-	log.Debugf("[ddl] txn %v fetches handle info %v takes time %v", txn.StartTS(), handleInfo, time.Since(startTime))
-	if ret.count == 0 {
-		return nil, ret
+	if isEnd {
+		ret.isAllDone = true
 	}
+	ret.count = len(w.idxRecords)
+	log.Debugf("[ddl] txn %v fetches handle info %v, ret %v, takes time %v", txn.StartTS(), w.taskRange, ret, time.Since(startTime))
 
-	err = d.getIndexRecords(t, taskOpInfo, rawRecords, idxRecords)
-	if err != nil {
-		ret.err = errors.Trace(err)
-	}
-	return idxRecords, ret
+	return w.idxRecords, ret
 }
 
-func (d *ddl) getIndexRecords(t table.Table, taskOpInfo *indexTaskOpInfo, rawRecords [][]byte, idxRecords []*indexRecord) error {
+func (w *worker) getIndexRecord(t table.Table, taskOpInfo *indexTaskOpInfo, rawRecord []byte, idxRecord *indexRecord) error {
 	cols := t.Cols()
-	ctx := d.newContext()
 	idxInfo := taskOpInfo.tblIndex.Meta()
-	defaultVals := make([]types.Datum, len(cols))
-	for i, idxRecord := range idxRecords {
-		rowMap, err := tablecodec.DecodeRow(rawRecords[i], taskOpInfo.colMap, time.UTC)
-		if err != nil {
-			return errors.Trace(err)
-		}
-		idxVal := make([]types.Datum, len(idxInfo.Columns))
-		for j, v := range idxInfo.Columns {
-			col := cols[v.Offset]
-			if col.IsPKHandleColumn(t.Meta()) {
-				if mysql.HasUnsignedFlag(col.Flag) {
-					idxVal[j].SetUint64(uint64(idxRecord.handle))
-				} else {
-					idxVal[j].SetInt64(idxRecord.handle)
-				}
-				continue
-			}
-			idxColumnVal := rowMap[col.ID]
-			if _, ok := rowMap[col.ID]; ok {
-				idxVal[j] = idxColumnVal
-				continue
-			}
-			idxColumnVal, err = tables.GetColDefaultValue(ctx, col, defaultVals)
-			if err != nil {
-				return errors.Trace(err)
+	err := tablecodec.DecodeRowWithMap(rawRecord, taskOpInfo.colMap, time.UTC, w.rowMap)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	idxVal := make([]types.Datum, len(idxInfo.Columns))
+	for j, v := range idxInfo.Columns {
+		col := cols[v.Offset]
+		if col.IsPKHandleColumn(t.Meta()) {
+			if mysql.HasUnsignedFlag(col.Flag) {
+				idxVal[j].SetUint64(uint64(idxRecord.handle))
+			} else {
+				idxVal[j].SetInt64(idxRecord.handle)
 			}
+			continue
+		}
+		idxColumnVal := w.rowMap[col.ID]
+		if _, ok := w.rowMap[col.ID]; ok {
 			idxVal[j] = idxColumnVal
+			continue
 		}
-		idxRecord.vals = idxVal
+		idxColumnVal, err = tables.GetColDefaultValue(w.ctx, col, w.defaultVals)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		idxVal[j] = idxColumnVal
 	}
+	idxRecord.vals = idxVal
 	return nil
 }
 
 const (
 	defaultBatchCnt      = 1024
 	defaultSmallBatchCnt = 128
 	defaultTaskHandleCnt = 128
```

Comment:
We use Seek to fetch handles and Seek is implement by Scan in TiKV, which has 256 batch size, change it to 256 may improve performance.